In [45]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import scipy.stats as st
import random
import seaborn as sns
import math
import scikit_posthocs as sp
from math import factorial as fact
from cmh import CMH
from scipy.stats import t
from scipy import stats
from scipy.stats import f_oneway

Задание 1

In [51]:
task_1 = pd.read_csv('Task_1_module2.csv')
task_1.head(5)


Time_point_1  Time_point_2
0       -0.7995        1.0018
1        1.0099       -4.1086
2        3.2986        1.6017
3       -4.5997       -5.0043
4        3.4326       -4.2815

In [50]:
len(task_1)

250

In [53]:
W_statistic, w_p_val = stats.shapiro(task_1['Time_point_1'])

print('W-statistic:', round(W_statistic, 3))
print('p-value:', round(w_p_val, 3))
if round(w_p_val, 3)>=0.05:
    print("Выборочное распределение подчиняется нормальному закону")
else:
    print("Выборочное распределение не подчиняется нормальному закону")

W-statistic: 0.997
p-value: 0.911
Выборочное распределение подчиняется нормальному закону


In [54]:
W_statistic, w_p_val = stats.shapiro(task_1['Time_point_2'])

print('W-statistic:', round(W_statistic, 3))
print('p-value:', round(w_p_val, 3))
if round(w_p_val, 3)>=0.05:
    print("Выборочное распределение подчиняется нормальному закону")
else:
    print("Выборочное распределение не подчиняется нормальному закону")

W-statistic: 0.993
p-value: 0.335
Выборочное распределение подчиняется нормальному закону


In [39]:
t_statistic, p_val = stats.ttest_rel(task_1['Time_point_1'], task_1['Time_point_2'])

print('t-statistic:', round(t_statistic, 3))
print('p-value:', round(p_val, 3))
if round(p_val, 3)<0.05:
    print("Различие статистически значимо")

t-statistic: -2.328
p-value: 0.021
Различие статистически значимо


Задание 2 

In [41]:
task_2 = pd.read_csv('Task_2_module2.csv')
task_2.head(5)

Time_point_1  Time_point_2
0        481.84        855.01
1        445.08         30.88
2        193.37         98.34
3         26.86        292.64
4         92.88         95.56

In [55]:
len(task_2)

250

In [56]:
W_statistic, w_p_val = stats.shapiro(task_2['Time_point_1'])

print('W-statistic:', round(W_statistic, 3))
print('p-value:', round(w_p_val, 3))
if round(w_p_val, 3)>=0.05:
    print("Выборочное распределение подчиняется нормальному закону")
else:
    print("Выборочное распределение не подчиняется нормальному закону")

W-statistic: 0.828
p-value: 0.0
Выборочное распределение не подчиняется нормальному закону


In [57]:
W_statistic, w_p_val = stats.shapiro(task_2['Time_point_2'])

print('W-statistic:', round(W_statistic, 3))
print('p-value:', round(w_p_val, 3))
if round(w_p_val, 3)>=0.05:
    print("Выборочное распределение подчиняется нормальному закону")
else:
    print("Выборочное распределение не подчиняется нормальному закону")

W-statistic: 0.837
p-value: 0.0
Выборочное распределение не подчиняется нормальному закону


 Т-критерий Вилкоксона — статистический критерий, используемый для проверки различий двух групп с непрерывными или порядковыми данными. Аналог t-критерия Стьюдента для зависимых выборок.

In [ ]:
scipy.stats.wilcoxon(x, y=None, zero_method='wilcox', correction=False, 
 alternative='two-sided', method='auto', *, axis=0, nan_policy='propagate', keepdims=False)[source]

In [42]:
t_statistic, p_val = stats.wilcoxon(task_2['Time_point_1'], task_2['Time_point_2'])

print('t-statistic:', round(t_statistic, 3))
print('p-value:', round(p_val, 3))
if round(p_val, 3)<0.05:
    print("Различие статистически значимо")
else:
    print("Различие статистически незначимо")    


t-statistic: 14751.0
p-value: 0.413
Различие статистически незначимо


Задание 3

In [43]:
task_3 = pd.read_csv('Task_3_module2.csv')
task_3.head(5)

Group_1  Group_2  Group_3  Group_4
0  -1.4944   7.0439  -7.4954   8.7465
1   0.3574   5.6963 -11.7362  10.1842
2   3.7011   6.7439 -14.1450  11.3717
3   1.0644   2.9705  -8.5661  11.0441
4  -6.4899   7.4093  -8.4214  12.8535

In [58]:
W_statistic, w_p_val = stats.shapiro(task_3['Group_1'])

print('W-statistic:', round(W_statistic, 3))
print('p-value:', round(w_p_val, 3))
if round(w_p_val, 3)>=0.05:
    print("Выборочное распределение подчиняется нормальному закону")
else:
    print("Выборочное распределение не подчиняется нормальному закону")

W-statistic: 0.996
p-value: 0.741
Выборочное распределение подчиняется нормальному закону


In [59]:
W_statistic, w_p_val = stats.shapiro(task_3['Group_2'])

print('W-statistic:', round(W_statistic, 3))
print('p-value:', round(w_p_val, 3))
if round(w_p_val, 3)>=0.05:
    print("Выборочное распределение подчиняется нормальному закону")
else:
    print("Выборочное распределение не подчиняется нормальному закону")

W-statistic: 0.99
p-value: 0.076
Выборочное распределение подчиняется нормальному закону


In [60]:
W_statistic, w_p_val = stats.shapiro(task_3['Group_3'])

print('W-statistic:', round(W_statistic, 3))
print('p-value:', round(w_p_val, 3))
if round(w_p_val, 3)>=0.05:
    print("Выборочное распределение подчиняется нормальному закону")
else:
    print("Выборочное распределение не подчиняется нормальному закону")

W-statistic: 0.995
p-value: 0.528
Выборочное распределение подчиняется нормальному закону


In [61]:
W_statistic, w_p_val = stats.shapiro(task_3['Group_4'])

print('W-statistic:', round(W_statistic, 3))
print('p-value:', round(w_p_val, 3))
if round(w_p_val, 3)>=0.05:
    print("Выборочное распределение подчиняется нормальному закону")
else:
    print("Выборочное распределение не подчиняется нормальному закону")

W-statistic: 0.996
p-value: 0.839
Выборочное распределение подчиняется нормальному закону


In [44]:
f_value, p_value = f_oneway(task_3['Group_1'], task_3['Group_2'], task_3['Group_3'], task_3['Group_4'])

print('statistic:', round(f_value, 3))
print('p-value:', round(p_value, 3))
if p_value<0.05:
    print ("Различие статистически значимо")
else:
    print("Различие статистически не значимо")

statistic: 2019.112
p-value: 0.0
Различие статистически значимо


Задание 4 

In [64]:
data_long = task_3.melt(var_name='Group', value_name='Value')
data_long.head(5)

Group   Value
0  Group_1 -1.4944
1  Group_1  0.3574
2  Group_1  3.7011
3  Group_1  1.0644
4  Group_1 -6.4899

In [66]:
import scikit_posthocs as sp
sp.posthoc_ttest(data_long, val_col='Value', group_col='Group', p_adjust="holm")

Group_1        Group_2        Group_3        Group_4
Group_1   1.000000e+00   1.763846e-63  1.037998e-146  1.340119e-146
Group_2   1.763846e-63   1.000000e+00  1.142708e-218   2.165798e-56
Group_3  1.037998e-146  1.142708e-218   1.000000e+00  8.126670e-270
Group_4  1.340119e-146   2.165798e-56  8.126670e-270   1.000000e+00

In [67]:
sp.posthoc_ttest(data_long, val_col='Value', group_col='Group', p_adjust="holm").round(3)

Group_1  Group_2  Group_3  Group_4
Group_1      1.0      0.0      0.0      0.0
Group_2      0.0      1.0      0.0      0.0
Group_3      0.0      0.0      1.0      0.0
Group_4      0.0      0.0      0.0      1.0

 Вывод: все группы между собой отличаются, то есть нет ни одной пары, которая была бы взята из одного распределения

Задание 5

In [23]:
task_5 = pd.read_csv('Task_5_module2.csv')
task_5.head(5)

Color    City
0    Синий   Токио
1   Жёлтый  Лондон
2   Жёлтый   Токио
3   Чёрный  Москва
4  Зелёный  Москва

In [68]:
len(task_5)

25000

In [36]:
crosstab_1 = pd.crosstab(task_5['City'], task_5['Color'])
crosstab_1

Color   Белый  Жёлтый  Зелёный  Красный  Синий  Сиреневый  Чёрный
City                                                             
Лондон   1183    1241     1114     1188   1182       1151    1156
Москва   1178    1180     1251     1233   1182       1200    1193
Токио    1224    1162     1223     1224   1185       1163    1187

Задание 6

In [74]:
chi2_task6, p_value_task6, dof_task6, expected_freq_task6 = st.chi2_contingency(crosstab_1)

In [73]:
print('Статистика равна:', round(chi2_task6,3))
print('P-value равен:', round(p_value_task6,3))
print('Число степеней свободы:', round(dof_task6,3))
print('Ожидаемые частоты:', np.round(expected_freq_task6,3))

Статистика равна: 12.818
P-value равен: 0.382
Число степеней свободы: 12
Ожидаемые частоты: [[1178.031 1177.374 1179.017 1197.747 1166.201 1154.7   1161.93 ]
 [1206.998 1206.324 1208.008 1227.199 1194.877 1183.094 1190.5  ]
 [1199.971 1199.302 1200.975 1220.054 1187.921 1176.206 1183.57 ]]


P-value = 0.382, не можем отвергнуть нулевую гипотезу об отсутствии зависимости между городом и цветом машины

In [77]:
print('Критическое значение равно:', round(st.chi2.ppf(0.95, df=dof_task6),3))

Критическое значение равно: 21.026


Статистика = 12.818, критическое значение равно 21.026. Статистика входит в область основной гипотезы. Не можем отвергнуть нулевую гипотезу об отсутствии зависимости между городом и цветом машины